openshift-knative-operator/pkg/monitoring/monitoring.go

```diff
@@ -21,45 +22,76 @@ import (
 )
 
 const (
-	EnableMonitoringEnvVar       = "ENABLE_SERVING_MONITORING_BY_DEFAULT"
-	EnableMonitoringLabel        = "openshift.io/cluster-monitoring"
-	ObservabilityCMName          = "observability"
-	ObservabilityBackendKey      = "metrics.backend-destination"
-	OpenshiftMonitoringNamespace = "openshift-monitoring"
-	prometheusRoleName           = "knative-serving-prometheus-k8s"
-	prometheusClusterRoleName    = "rbac-proxy-metrics-prom"
-	servingSMRbacManifestPath    = "SERVING_SERVICE_MONITOR_RBAC_MANIFEST_PATH"
+	EnableServingMonitoringEnvVar  = "ENABLE_SERVING_MONITORING_BY_DEFAULT"
+	EnableEventingMonitoringEnvVar = "ENABLE_EVENTING_MONITORING_BY_DEFAULT"
+	EnableMonitoringLabel          = "openshift.io/cluster-monitoring"
+	ObservabilityCMName            = "observability"
+	ObservabilityBackendKey        = "metrics.backend-destination"
+	OpenshiftMonitoringNamespace   = "openshift-monitoring"
+	prometheusRoleName             = "knative-prometheus-k8s"
+	prometheusClusterRoleName      = "rbac-proxy-metrics-prom"
+	smRbacManifestPath             = "SERVICE_MONITOR_RBAC_MANIFEST_PATH"
+)
+
+type KComponentType string
+
+const (
+	Serving  KComponentType = "Serving"
+	Eventing KComponentType = "Eventing"
 )
 
 var (
-	servingComponents = sets.NewString("activator", "autoscaler", "autoscaler-hpa", "controller", "domain-mapping", "domainmapping-webhook", "webhook")
+	servingComponents  = sets.NewString("activator", "autoscaler", "autoscaler-hpa", "controller", "domain-mapping", "domainmapping-webhook", "webhook")
+	eventingComponents = sets.NewString("eventing-controller", "eventing-webhook", "imc-controller", "imc-dispatcher", "mt-broker-controller", "mt-broker-filter", "mt-broker-ingress", "sugar-controller")
 )
 
 func init() {
 	_ = monitoringv1.AddToScheme(scheme.Scheme)
 }
 
-func ReconcileServingMonitoring(ctx context.Context, api kubernetes.Interface, ks *v1alpha1.KnativeServing) error {
-	backend := ks.Spec.CommonSpec.Config[ObservabilityCMName][ObservabilityBackendKey]
-	if shouldEnableMonitoring(backend) {
-		if err := reconcileMonitoringLabelOnNamespace(ctx, ks.Namespace, api, true); err != nil {
+func ReconcileMonitoringForKComponent(ctx context.Context, api kubernetes.Interface, comp v1alpha1.KComponent) error {
```

Comment:
This is imo unnecessary "code-deduplication". When calling this, you know if you want to deal with Serving and Eventing. You forget that through the boundary of this interface only to then pick it apart in the implementation after the interface again.

Can we come up with a version for Serving and Eventing that composes the "helper" functions herein but doesn't need `KComponentType`?

Reply:
yeah. would be good to call into specific code for each "Component", from the monitoring file

Reply:
I made it a bit agnostic. If I do it the way you propose (initially it was like this) then it means I need to calculate several parameters before `ReconcileMonitoringForKComponent` is called like the common spec, ns all at the extension side so I can take advantage that I know what instance the comp is and avoid type casting. Also I would have to pass the name of the instance explicitly because it is used in logging. It is a trade-off, either pollute the extension code and use the info or make it less specific and hide the details.

Reply:
I guess the point is that it doesn't quite hide the details, but just moves them to other places. The monitoring functions "fake" a genericness that doesn't seem to be applicable.

Reply:
Extension code does not know the details. That is what I mean by hiding.